### PR TITLE
Initial refactoring to use fishcakez/connection instead of sync connect

### LINKED ISSUE #11: Connection handling
Currently connection establishment is synchronous and badly handled. If a broker is unavailable the startup process just hangs.

Possibly investigate the use of the Connection library to implement a robust async connection with backoff mechanism.

### PR DESCRIPTION
**Pull request by lucacorti**
This should fix #11

### CODE REVIEW

**Comment by lucacorti** [comment]
@lswith Can you test with the connection branch if this improves your use case at all?

**Comment by lswith** [comment]
I'll take a look but my problem is mainly around bad configuration. If I supply bad configuration, the startup doesn't happen. I'll see if this changes anything.

**Comment by lucacorti** [comment]
If you can provide a test case with a failing configuration I can tackle that in as a separate issue.

**Comment by lucacorti** [comment]
Check out `test/lapin_test.exs` which contains all the machinery to do this.

**Comment by lswith** [comment]
https://github.com/lswith/lapin/blob/master/test/lapin_test.exs#L87

**Comment by lucacorti** [comment]
Can you setup Travis CI integration for your repo and put the test in the connection branch so we can see the test run on Travis directly. Once you connect your repo, travis will pickup the .travis.yml and automatically run tests.

**Comment by lswith** [comment]
https://travis-ci.org/lswith/lapin/builds/306045661

**Comment by lucacorti** [comment]
Ok, so I think we are conflating a number of unrelated issues here.

What I did on the `connection` branch was basically swap out the sync connect on `Lapin.Connection.start_link` for an async connect provided via the `Connection` library. This addresses the issue of application startup, since the connect is not blocking application startup anymore anymore. However, this also means we are not returning errors on process startup. So your issue with the boot hanging indefinitely should now be gone, but you are not receiving any feedback on the broker connection.

Currently, everything you put in the configuration is started (and restarted on crash) by the `Lapin.Connection.Supervisor` which starts each `Lapin.Connection` process. The processes now run regardless of the underlying broker connection state. I guess this is good under supervision, because there is no way to tell if the connect will always be failing because of bad configuration or if will eventually succeed. Some kind of logging should be provided to signal when connections are actually established/retried and some backoff algorithm should be implemented to avoid constant retries. On publishing, errors should be returned if the broker connection is not established yet so callers can retry. 

Additionally, a sync startup option could be provided to ensure `Lapin.Connection` processes started manually out of the supervision three either establish broker connections on startup or error out.

I'll start working on these mechanisms. Meanwhile you can change your test by removing the assert_raise which will never be raised, and that could stay as a test of async connection startup.

**Comment by lucacorti** [comment]
So, my last commit adds error reporting on publish. I've updated the bad host test for this use case.
You'll also notice that test makes use of AMQP uri in the configuration instead of individual settings, also addressing #3. I'm going to work on reconnection backoff (which is currently hardcoded to 1 second) and sync connect next.

**Comment by lswith** [comment]
this definitely solved the startup issues. Just had a test today and it's perfect. I'm looking forward to the new release.

File: lib/lapin/connection.ex [hunk 1]
```diff
@@ -352,6 +360,42 @@ defmodule Lapin.Connection do
...
+
+  defp map_uri(%URI{} = uri) do
+    uri
+    |> Map.from_struct()
+    |> Enum.to_list()
+    |> uri_to_list()
+  end
+
+  defp uri_to_list(uri) when is_list(uri) do
+    with {vhost, uri} <- Keyword.pop(uri, :path),
```
Comment (lswith) [inline lib/lapin/connection.ex]: this vhost isn't quite correct. It needs to remove the leading "/" to connect correctly. Something like:
```elixir
vhost = String.replace_leading(vhost, "/", "")				
vhost =		
   case vhost do		
     "" -> "/"		
     _ -> vhost		
   end
```

Comment (lucacorti) [reply lib/lapin/connection.ex]: Thanks, latest commit should fix this.

**Pull request merged**